apps/common/lib/lexical/ast/analysis/analyzer.ex

```diff
@@ -0,0 +1,424 @@
+defmodule Lexical.Ast.Analysis.Analyzer do
+  @moduledoc false
+
+  alias __MODULE__
+  alias Lexical.Document
+  alias Lexical.Document.Position
+  alias Lexical.Document.Range
+  alias Sourceror.Zipper
+
+  @scope_id :_scope_id
+
+  @block_keywords [:do, :else, :rescue, :catch, :after]
+  @clauses [:->]
+
+  defmodule Alias do
+    defstruct [:module, :as, :line]
+
+    def new(module, as, line) when is_list(module) and is_atom(as) and line > 0 do
+      %Alias{module: module, as: as, line: line}
+    end
+  end
+
+  defmodule Scope do
+    defstruct [:id, :range, module: [], parent_aliases: %{}, aliases: []]
+
+    def new(id, %Range{} = range, parent_aliases \\ %{}, module \\ []) do
+      %Scope{id: id, range: range, module: module, parent_aliases: parent_aliases}
+    end
+
+    def global do
+      %Scope{id: :global}
+    end
+
+    def alias_map(%Scope{} = scope) do
+      scope.aliases
+      |> Enum.sort_by(& &1.line)
+      |> Map.new(&{&1.as, &1})
+      |> Enum.into(scope.parent_aliases)
+    end
+
+    def empty?(%Scope{aliases: []}), do: true
```

Comment:
nit: scopes are about more than just aliases. Imports, function definitions, and variables will all live here soon.

Reply:
You're right, but that's why this is a function in the `Scope` module instead of an ad-hoc check for empty `scope.aliases` somewhere else. We'll indeed have to update this as scopes start to store more information. (I want to start aggregating imports as soon as possible so that we can start resolving unqualified calls in `Entity.resolve`.)